### PR TITLE
Adds testing to abstract llm and simplifies fallback handling

### PR DESCRIPTION
**Pull request by vmesel**

### CODE REVIEW

**Review by avelino — COMMENTED** [review]
remove `fallback_not_found_relevant_contents`?

**Comment by vmesel** [comment]
> remove fallback_not_found_relevant_contents?

@avelino IMO, it doesn't make sense to have this variable, it's only being used in the AbstractLLM (which is never used) and you could just use the standard `fallback` as the field, just overwritting the behavior of the DefaultLLM or adding a custom instance of your own.

File: src/dialog/llm/abstract_llm.py [hunk 1]
```diff
@@ -79,9 +79,11 @@ def process(self, input: str):
         """
         processed_input = self.preprocess(input)
         self.generate_prompt(processed_input)
+
         if len(self.relevant_contents) == 0 and \
-            self.config.get("prompt").get("fallback_not_found_relevant_contents"):
-            return {"text": self.config.get("prompt").get("fallback_not_found_relevant_contents")}
+            self.config.get("prompt").get("fallback"):
+            return {"text": self.config.get("prompt").get("fallback")}
+
```
Comment (vmesel) [inline src/dialog/llm/abstract_llm.py:86]: @avelino as you can see here, if we have an if statement that doesn't use the LLM inside the `process(self, input)` function, we are good

File: Dockerfile [hunk 2]
```diff
@@ -11,7 +11,7 @@ ENV PYTHONUNBUFFERED=1
 
 WORKDIR /app
 
-COPY poetry.lock pyproject.toml /app/
+COPY poetry.lock pyproject.toml README.md /app/
```
[the remark this answers is not in the document]
Comment (vmesel) [reply Dockerfile:14]: Unfortunately it is so we can prevent a future error in poetry.

```
dialog-1  | Skipping virtualenv creation, as specified in config file.
dialog-1  | Installing dependencies from lock file
dialog-1  |
dialog-1  | Package operations: 22 installs, 0 updates, 0 removals
dialog-1  |
dialog-1  |   - Installing asttokens (2.4.1)
dialog-1  |   - Installing executing (2.0.1)
dialog-1  |   - Installing parso (0.8.3)
dialog-1  |   - Installing pure-eval (0.2.2)
dialog-1  |   - Installing traitlets (5.14.1)
dialog-1  |   - Installing wcwidth (0.2.13)
dialog-1  |   - Installing decorator (5.1.1)
dialog-1  |   - Installing jedi (0.19.1)
dialog-1  |   - Installing matplotlib-inline (0.1.6)
dialog-1  |   - Installing pluggy (1.4.0)
dialog-1  |   - Installing prompt-toolkit (3.0.43)
dialog-1  |   - Installing pygments (2.17.2)
dialog-1  |   - Installing stack-data (0.6.3)
dialog-1  |   - Installing ipython (8.21.0)
dialog-1  |   - Installing pathspec (0.12.1)
dialog-1  |   - Installing pytest (8.1.1)
dialog-1  |   - Installing black (23.12.1)
dialog-1  |   - Installing ipdb (0.13.13)
dialog-1  |   - Installing isort (5.13.2)
dialog-1  |   - Installing pytest-alembic (0.11.0)
dialog-1  |   - Installing pytest-mock (3.12.0)
dialog-1  |   - Installing responses (0.25.0)
dialog-1  |
dialog-1  | Installing the current project: dialog (0.1.0)
dialog-1  | Warning: The current project could not be installed: [Errno 2] No such file or directory: '/app/README.md'
dialog-1  | If you do not want to install the current project use --no-root.
dialog-1  | If you want to use Poetry only for dependency management but not for packaging, you can disable package mode by setting package-mode = false in your pyproject.toml file.
```

Command to replicate:

```
make test-build
```

**Comment by lgabs** [comment]
> @avelino IMO, it doesn't make sense to have this variable, it's only being used in the AbstractLLM (which is never used) and you could just use the standard fallback as the field, just overwriting the behavior of the DefaultLLM or adding a custom instance of your own.

In [this issue](https://github.com/talkdai/dialog/issues/117) we discussed some ideas for fallback, and @avelino proposed the idea of using two fallbacks in case no relevant documents were found from the retriever:

1. A fixed fallback text as an answer
2. Another prompt that is injected in prompt such as we delegate the answer to the llm (which, depending of prompt instructions, will answer accordingly) 

The idea of his PR was that. Should we change this behavior?

obs: I think the terms were mixed in the code, because the fixed text here
https://github.com/talkdai/dialog/blob/4ca50e80ea3968e467c133df688bc62f5cd17aa1/src/dialog/llm/abstract_llm.py#L84

is using `fallback_not_found_relevant_contents` while this other part

https://github.com/talkdai/dialog/blob/4ca50e80ea3968e467c133df688bc62f5cd17aa1/src/dialog/llm/default.py#L53

uses `prompt_config.get("fallback") or self.config.get("fallback").get("prompt")`.

**Comment by vmesel** [comment]
@lgabs the thing is, abstractllm is never called since its an abstract class, so this fallback is never used, unless you've built a new LLM that relies on this `fallback_not_found_relevant_contents`. We can do a rollout with the variables still inside the settings, but there is no point of maintaining this variable inside the AbstractLLM.

This variable should be deprecated in the future when we get to a better understanding on how to process fallbacks on no retrievals.

**Comment by vmesel** [comment]
In the issue, you made this comment @lgabs, which I agree, but my point here is to maintain a single variable that carries that data, my choice was to keep fallback since its shorter and simple.

> Probably the default behavior should be the fixed fallback, and if the developer wants to delegate to GPT the answer using instructions from the prompt template, than he/she can make its own custom subclass of AbstractLLM dealing with desired behavior.

**Comment by vmesel** [comment]
I'm going to fallback this PR to just add the test to the AbstractLLM, later we discuss the fallback variable

**Comment by lgabs** [comment]
@vmesel 
> abstractllm is never called since its an abstract class, so this fallback is never used

this is not true, the default `DialogLLM` [here](https://github.com/talkdai/dialog/blob/4ca50e80ea3968e467c133df688bc62f5cd17aa1/src/dialog/llm/default.py#L20) is a subclass of the abstract one and it does not re-implement the abstrac's method `process`, and indeed it is called here (for any new subclass made):

https://github.com/talkdai/dialog/blob/4ca50e80ea3968e467c133df688bc62f5cd17aa1/src/main.py#L83

to avoid leaving this behavior in the abstract, the method would be re-implemented in default class, but this would be strange since the default class should probably work with minimal changes from the abstract. 

We could reopen the issue to discuss that later.

**Pull request closed without merging**